File: csharp/test/Vista.SDK.Tests/ISOStringTests.cs
namespace Vista.SDK.Tests;

public class IsISOStringTests
{
    public record SmokeContext
    {
        public int Count { get; set; } = 0;
        public int Succeeded { get; set; } = 0;
        public List<(string LocalId, Exception? e)> Errors { get; } = new();
    }

    const string AllAllowedCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~";

    public static IEnumerable<object[]> Valid_Characters =>
        AllAllowedCharacters.Split("", StringSplitOptions.RemoveEmptyEntries).Select(c => new object[] { c[0] });
    public static IEnumerable<object[]> Spot_Test_Data =>

        [
            ["test", true],
            ["TeST", true],
            ["with space", false],
            ["#%/*", false],
        ];

    [Theory]
    [MemberData(nameof(Valid_Characters))]
    public void AllValidCharacters(char c) => Assert.True(VIS.IsISOString(c));

    [Fact]
    public void AllAllowedInOne() => Assert.True(VIS.IsISOString(AllAllowedCharacters));

    [Theory]
    [MemberData(nameof(Spot_Test_Data))]
    public void SpotTests(string input, bool expected) => Assert.Equal(expected, VIS.IsISOString(input));

    [Fact]
    public async Task SmokeTest_Parsing()
    {
        await using (var file = File.OpenRead("testdata/LocalIds.txt"))
        {
            using var reader = new StreamReader(file, leaveOpen: true);

            var context = new SmokeContext();

            string? localIdStr;

            while ((localIdStr = await reader.ReadLineAsync()) is not null)
            {
                try
                {
                    var match = VIS.MatchISOLocalIdString(localIdStr);
                    if (!match)
                        context.Errors.Add((localIdStr, null));
                    else
                    {
                        context.Succeeded++;
                    }
                }
                catch (Exception ex)
                {
                    context.Errors.Add((localIdStr, ex));
                }
                context.Count++;
            }

            if (context.Errors.Any())
            {
                context
                    .Errors
                    .ForEach(e =>
                    {
                        Console.WriteLine($"Failed to parse {e.LocalId} with error {e.e?.Message ?? "Not a match"}");
                    });
            }

            Assert.Empty(context.Errors);
            Assert.Equal(context.Succeeded, context.Count);
        }
    }
}

File: cpp/test/ISOStringTests.cpp
/**
 * @file ISOStringTests.cpp
 * @brief Unit tests for ISO string validation utilities.
 */

#include "pch.h"

#include "dnv/vista/sdk/VIS.h"

namespace dnv::vista::sdk
{
	struct SmokeContext
	{
		int count = 0;
		int succeeded = 0;
		std::vector<std::pair<std::string, std::optional<std::string>>> errors;
	};

	const std::string AllAllowedCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~";

	TEST( IsISOStringTests, AllValidCharacters )
	{
		for ( const char ch : AllAllowedCharacters )
		{
			EXPECT_TRUE( VIS::isISOString( std::string( 1, ch ) ) )
				<< "Character: " << ch;
		}
	}

	TEST( IsISOStringTests, AllAllowedInOne )
	{
		EXPECT_TRUE( VIS::isISOString( AllAllowedCharacters ) );
	}

	TEST( IsISOStringTests, SpotTests )
	{
		struct TestCase
		{
			std::string input;
			bool expectedResult;
		};

		std::vector<TestCase> testCases = {
			{ "test", true },
			{ "TeST", true },
			{ "with space", false },
			{ "#%/*", false },
		};

		for ( const auto& testCase : testCases )
		{
			EXPECT_EQ( VIS::isISOString( testCase.input ), testCase.expectedResult ) << "Input: " << testCase.input;
		}
	}

	TEST( IsISOStringTests, SmokeTest_Parsing )
	{
		std::vector<std::string> possiblePaths = {
			"testdata/LocalIds.txt",
			"../testdata/LocalIds.txt",
			"../../testdata/LocalIds.txt",
			"../../../testdata/LocalIds.txt",
		};

		std::ifstream file;
		std::string attemptedPaths;
		bool fileOpened = false;

		for ( const auto& path : possiblePaths )
		{
			file.open( path );
			if ( file.is_open() )
			{
				SPDLOG_INFO( "Found test data at path: {}", path );
				fileOpened = true;
				break;
			}

			attemptedPaths += path + ", ";
			file.clear();
			SPDLOG_DEBUG( "Failed to open test data file: {}", path );
		}

		if ( !fileOpened )
		{
			SPDLOG_ERROR( "Failed to open test data file. Attempted paths: {}", attemptedPaths );
			ASSERT_TRUE( false ) << "Failed to open LocalIds.txt. Attempted paths: " << attemptedPaths;
			return;
		}

		SmokeContext context;

		std::string localIdStr;
		while ( std::getline( file, localIdStr ) )
		{
			try
			{
				bool match = VIS::matchISOLocalIdString( localIdStr );
				if ( !match )
				{
					context.errors.push_back( { localIdStr, std::nullopt } );
				}
				else
				{
					context.succeeded++;
				}
			}
			catch ( const std::exception& ex )
			{
				context.errors.push_back( { localIdStr, ex.what() } );
			}
			context.count++;
		}

		if ( !context.errors.empty() )
		{
			for ( [[maybe_unused]] const auto& [localId, error] : context.errors )
			{
				SPDLOG_ERROR( "Failed to parse {} with error {}", localId, error.has_value() ? *error : "Not a match" );
			}
		}

		EXPECT_TRUE( context.errors.empty() ) << "Found " << context.errors.size() << " errors";
		EXPECT_EQ( context.count, context.succeeded );
	}
}
